From dfb114e46290ba5652ad9f3601b3eb519bb2a33e Mon Sep 17 00:00:00 2001
From: Doug Smith <doug.smith@ezurio.com>
Date: Fri, 16 Feb 2024 00:56:18 -0500
Subject: [PATCH] FIPS: fetch entropy in fips module

Remove drbg chaining in fips mode/module
so that entropy will be fetched in
the fips module.
---
 crypto/rand/rand_lib.c | 28 +++++++++++++++++++++-------
 1 file changed, 21 insertions(+), 7 deletions(-)

diff --git a/crypto/rand/rand_lib.c b/crypto/rand/rand_lib.c
index 14999540ab..f3c359bc71 100644
--- a/crypto/rand/rand_lib.c
+++ b/crypto/rand/rand_lib.c
@@ -778,7 +778,7 @@ EVP_RAND_CTX *rand_get0_primary(OSSL_LIB_CTX *ctx)
 
 #if !defined(FIPS_MODULE) || !defined(OPENSSL_NO_FIPS_JITTER)
     /* Create a seed source for libcrypto or jitter enabled FIPS provider */
-    if (seed == NULL) {
+    if (seed == NULL && !EVP_default_properties_is_fips_enabled(ctx)) {
         ERR_set_mark();
         seed = newseed = rand_new_seed(ctx);
         ERR_pop_to_mark();
@@ -847,9 +847,15 @@ EVP_RAND_CTX *rand_get0_public(OSSL_LIB_CTX *ctx)
 
     rand = CRYPTO_THREAD_get_local(&dgbl->public);
     if (rand == NULL) {
-        primary = rand_get0_primary(ctx, dgbl);
-        if (primary == NULL)
-            return NULL;
+        primary = NULL;
+#ifndef FIPS_MODULE
+        if (!EVP_default_properties_is_fips_enabled(ctx)) {
+            primary = rand_get0_primary(ctx, dgbl);
+            if (primary == NULL)
+                return NULL;
+        }
+        // note: FIPS module fetches entropy directly
+#endif
 
         ctx = ossl_lib_ctx_get_concrete(ctx);
 
@@ -886,9 +892,15 @@ EVP_RAND_CTX *rand_get0_private(OSSL_LIB_CTX *ctx)
 
     rand = CRYPTO_THREAD_get_local(&dgbl->private);
     if (rand == NULL) {
-        primary = rand_get0_primary(ctx, dgbl);
-        if (primary == NULL)
-            return NULL;
+        primary = NULL;
+#ifndef FIPS_MODULE
+        if (!EVP_default_properties_is_fips_enabled(ctx)) {
+            primary = rand_get0_primary(ctx, dgbl);
+            if (primary == NULL)
+                return NULL;
+        }
+        // note: FIPS module fetches entropy directly
+#endif
 
         ctx = ossl_lib_ctx_get_concrete(ctx);
 
-- 
2.34.1
